crates/sui/tests/transaction_orchestrator_tests.rs

```diff
@@ -137,6 +154,116 @@ async fn test_fullnode_wal_log() -> Result<(), anyhow::Error> {
+    }
+
+    // Wait for all nodes to reach the next epoch.
+    wait_for_nodes_transition_to_epoch(authorities.iter().chain(fullnodes.iter()), 1).await;
+
+    // Give it some time for the update to happen
+    tokio::time::sleep(tokio::time::Duration::from_secs(3)).await;
+    fullnode.with(|node| {
+        let epoch = node
+            .transaction_orchestrator()
+            .unwrap()
+            .quorum_driver()
+            .current_epoch();
+        assert_eq!(epoch, 1);
+        assert_eq!(
+            node.clone_authority_aggregator().unwrap().committee.epoch,
+            1
+        );
+    });
+}
+
+#[tokio::test]
+async fn test_tx_across_epoch_boundaries() {
+    telemetry_subscribers::init_for_testing();
+    let (sender, keypair) = get_key_pair::<AccountKeyPair>();
+    let gas_objects = generate_test_gas_objects_with_owner(180, sender);
+
+    let (authorities, mut fullnodes) =
+        spawn_test_authorities_with_fullnodes(gas_objects.clone(), &test_authority_configs(), 1)
+            .await;
+    let txes = gas_objects
+        .iter()
+        .map(|o| {
+            let data = TransactionData::new_transfer_sui(
+                get_key_pair::<AccountKeyPair>().0,
+                sender,
+                None,
+                o.compute_object_reference(),
+                5000,
+            );
+            to_sender_signed_transaction(data, &keypair)
+        })
+        .collect::<Vec<_>>();
+    let fullnode = fullnodes.swap_remove(0);
+    let to = fullnode.with(|node| node.transaction_orchestrator().unwrap());
+
+    let handle = tokio::task::spawn(async move {
+        for tx in txes {
+            let tx_digest = *tx.digest();
+            info!(?tx_digest, "Submitting tx");
+            let tx = tx.into_inner();
+            loop {
+                match to
+                    .execute_transaction(ExecuteTransactionRequest {
+                        transaction: tx.clone(),
+                        request_type: ExecuteTransactionRequestType::WaitForEffectsCert,
+                    })
+                    .await
+                {
+                    Ok(_) => {
+                        info!(?tx_digest, "tx result: ok");
+                        break;
+                    }
+                    Err(QuorumDriverError::TimeoutBeforeReachFinality) => {
+                        info!(?tx_digest, "tx result: timeout and will retry")
+                    }
+                    Err(other) => panic!("unexpected error: {:?}", other),
+                }
+                // Retry a transaction if it times out (e.g. in epoch boundary) after 500 ms.
+                tokio::time::sleep(tokio::time::Duration::from_millis(500)).await;
+            }
+            // Wait for a little bit before getting to the next tx.
+            tokio::time::sleep(tokio::time::Duration::from_millis(100)).await;
+        }
+    });
+    
+
+    // uncomment this will cause fullnode not be able to advance epoch
```

Comment:
^^

Reply:
Is this a bug tracked somewhere?

Reply:
@mystenmark  and @williampsmith have been looking into it

Reply:
Out of curiosity, does this bug remain after rebasing now? There were 2 semi-related bug fixes recently merged